File: javargarten/src/irrgarten/Shield.java
package irrgarten;

public class Shield {
/*Esta clase representa los escudos que utiliza el jugador cuando se defiende de 
un ataque de un monstruo.*/

    private float protection;
    private int uses;



    /**************************************************************************/

    /*Constructor por defecto: crea un escudo sin protección ni usos*/
    public Shield(){this(Dice.shieldPower(), Dice.MAX_USES());}
    /*Constructor con especificación de cada parámetro */
    public Shield(float protection, int uses){
        this.protection = protection;
        this.uses = uses;
    }
    /**************************************************************************/


    /*Devuelve  la intensidad de la defensa del jugador y resta un uso al escudo 
    si éste puede proteger (uses>0). */
    public float protect(){
        if (uses > 0){
            uses--;
            return protection;
        }
        return 0;
    }

    /*Decide si un escudo debe ser descartado con ayuda del Dado. */
    public boolean discard(){ return Dice.discardElement(uses); }

    //Devuelve una String del formato 'S[{power}, {uses}]'
    @Override
    public String toString(){ return "S[" + protection + ", " + uses + "]"; }
}

File: rubyrrgarten/Shield.rb
module Irrgarten
  class Shield
    # Atributos
    attr_reader :protection, :uses

    # Constructor
    def initialize(protection, uses)
      @protection = protection
      @uses = uses
    end

    # Método para realizar la protección
    def protect
      if @uses > 0
        @uses -= 1  # Decrementa el uso
        return @protection  # Devuelve el valor de protección
      end
      0  # Devuelve 0 si no hay usos disponibles
    end

    # Método para decidir si el escudo debe ser descartado
    def discard
      Dice.discard_element(@uses)  # Llama al método de la clase Dice
    end

    # Método para representar el escudo como un string
    def to_s
      "S[#{@protection}, #{@uses}]"  # Formato 'S[{protection}, {uses}]'
    end
  end
end
